llext: add a section reordering script

laurenmurphyx64 opened the pull request:
UPDATE: This PR has replaced the custom Python script for reordering with a linker script instead.

---

In some cases, the compiler generates objects that have sections "out of order" for the LLEXT loader, i.e. regions overlap. Reordering variable declarations, code, header includes, etc. rarely fixes the issue. This script detects if the LLEXT loader will reject an ET_REL object and reorders the sections to match LLEXT regions.

Originally included in https://github.com/zephyrproject-rtos/zephyr/pull/93627.

Below is a list of test extensions that the compiler generated with sections in an invalid order.

    syscalls for x86 platforms with MMU (can be corrected by changing header order)
    init_fini for nsim/nsim_vpx5 with arcmwdt (can't be corrected)
    All extensions on qemu_malta*
    All extensions on qemu_leon3/leon3*

*We don't support these architectures yet, and currently just build at the moment, but the extensions as-is would fail to load if we tried

A key point to note about the reordering script is that it would require users using detached sections to name the section in the form ".detached.*", because there is no other way for a Python script invoked from CMake to detect a section is detached without it and forego reordering.

Another key point to note: I looked into the suggestion of replacing this script with an extra step calling the linker with a linker script. I mistakenly thought this wasn't possible with ET_REL objects, but it is, theoretically. For example:

```
$ readelf -S syscalls.llext.bak 
There are 14 section headers, starting at offset 0x430:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        00000000 000034 0000f2 00  AX  0   0  1
  [ 2] .rel.text         REL             00000000 00033c 000078 08   I 11   1  4
  [ 3] .data             PROGBITS        00000000 000126 000000 00  WA  0   0  1
  [ 4] .bss              NOBITS          00000000 000128 000008 00  WA  0   0  4
  [ 5] .rodata           PROGBITS        00000000 000128 0000ef 00   A  0   0  4
  [ 6] .pinned_text      PROGBITS        00000000 000217 000012 00  AX  0   0  1
  [ 7] .rel.pinned_text  REL             00000000 0003b4 000008 08   I 11   6  4
  [ 8] .exported_sym     PROGBITS        00000000 00022c 000008 00   A  0   0  4
  [ 9] .rel.exported_sym REL             00000000 0003bc 000010 08   I 11   8  4
  [10] .comment          PROGBITS        00000000 000234 000021 01  MS  0   0  1
  [11] .symtab           SYMTAB          00000000 000258 000090 10     12   3  4
  [12] .strtab           STRTAB          00000000 0002e8 000052 00      0   0  1
  [13] .shstrtab         STRTAB          00000000 0003cc 000064 00      0   0  1
...
...

$ cat link.ld 
SECTIONS {
        .pinned_text : {
                *(.pinned_text .pinned_text.*)
        }

        .text : {
                *(.text .text.*)
        }
}

$ /home/laurenmu/zephyr-sdk-0.17.0/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -m32 -r -T link.ld syscalls.llex
t.bak -o syscalls.llext.reorder

$ readelf -S syscalls.llext.reorder 
There are 14 section headers, starting at offset 0x47c:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .pinned_text      PROGBITS        00000000 000034 000012 00  AX  0   0  1
  [ 2] .rel.pinned_text  REL             00000000 000388 000008 08   I 11   1  4
  [ 3] .text             PROGBITS        00000012 000046 0000f2 00  AX  0   0  1
  [ 4] .rel.text         REL             00000000 000390 000078 08   I 11   3  4
  [ 5] .rodata           PROGBITS        00000000 000138 0000ef 00   A  0   0  4
  [ 6] .exported_sym     PROGBITS        00000000 000228 000008 00   A  0   0  4
  [ 7] .rel.exported_sym REL             00000000 000408 000010 08   I 11   6  4
  [ 8] .data             PROGBITS        00000000 000230 000000 00  WA  0   0  1
  [ 9] .bss              NOBITS          00000000 000230 000008 00  WA  0   0  4
  [10] .comment          PROGBITS        00000000 000230 000021 01  MS  0   0  1
  [11] .symtab           SYMTAB          00000000 000254 0000e0 10     12   8  4
  [12] .strtab           STRTAB          00000000 000334 000052 00      0   0  1
  [13] .shstrtab         STRTAB          00000000 000418 000064 00      0   0  1
...
```

One issue is that various architectures, toolchains, Zephyr features, and users may generate custom sections with names that don't match the usual linker section names. The second is that in relocatable mode (-r), the linker merges sections of the same name. For example, I can't match the individual .text.subname sections to order them without combining them into one section. This might be undesirable. I don't know. 

I will try implementing a linker script that works with x86's MMU's pinned_text at least, maybe fix NSIM ARCMWDT's init_fini issue too once the other PR (link above) is merged. I'm submitting this PR now mostly to run the script through SonarQube again.

laurenmurphyx64 commented:
A lot simpler to use the linker script to reorder sections. What do we think?

laurenmurphyx64 commented:
Am I actually able to retry the analysis without being an admin? I don't see an option in SonarQube

teburd commented:
> Am I actually able to retry the analysis without being an admin? I don't see an option in SonarQube

@kartben maybe knows, I think the basic way is closing and reopening the same PR?

lgirdwood commented:
Adding @lyakh @serhiy-katsyuba-intel for comments.

teburd commented:
@tejlmand can you take look, been a few months

tejlmand reviewed (changes requested):
Mostly minor nits.

Commit message states:
> Enables compilation on MIPS and native_sim
so section reordering can be tested by loader.

In that case, please update current llext test cases so that we can verify the improved behavior for those platforms.

laurenmurphyx64 commented on arch/posix/CMakeLists.txt:
```diff
@@ -50,13 +50,17 @@ else() # Linux.x86_64
     target_link_options(native_simulator INTERFACE "-m32")
     target_compile_options(native_simulator INTERFACE "-m32")
 
+    set(LLEXT_APPEND_FLAGS ${LLEXT_APPEND_FLAGS} -m32 CACHE STRING "LLEXT append flags")
```
(See above)

laurenmurphyx64 commented on cmake/modules/extensions.cmake:
```diff
@@ -5982,6 +5982,25 @@ function(add_llext_target target_name)
+    -r ${LLEXT_APPEND_FLAGS}
+    ${TOPT} ${ZEPHYR_BASE}/subsys/llext/llext_reorder_sections.ld
+    ${llext_pkg_output}.bak -o ${llext_pkg_output}
+  )
```
I think I fixed it with this push. The viewer on GitHub.com is a little confusing. Let me know if I didn't

laurenmurphyx64 commented on cmake/modules/extensions.cmake:
```diff
@@ -5982,6 +5982,25 @@ function(add_llext_target target_name)
+  # reorder the sections in these cases
+  if (CONFIG_LLEXT_TYPE_ELF_OBJECT)
+  set(reorder_sects_cmd
+    ${CMAKE_COMMAND} -E copy ${llext_pkg_output} ${llext_pkg_output}.bak &&
```
Split into two

laurenmurphyx64 commented:
> Mostly minor nits.
> 
> Commit message states:
> 
> > Enables compilation on MIPS and native_sim
> > so section reordering can be tested by loader.
> 
> In that case, please update current llext test cases so that we can verify the improved behavior for those platforms.

@tejlmand I may have misused the word "tested" there, sorry. They compile and run successfully through the region allocation step that was failing before, but linking step after that still fails for all tests on those platforms, so if I added them the tests would still fail. Do you want me to write a test that calls the loading functions up to llext_map_sections?

tejlmand commented on cmake/compiler/gcc/target_mips.cmake:
```diff
@@ -6,9 +6,11 @@
 if(CONFIG_BIG_ENDIAN)
   list(APPEND TOOLCHAIN_C_FLAGS -EB)
   list(APPEND TOOLCHAIN_LD_FLAGS -EB)
+  set(LLEXT_APPEND_FLAGS -EB)
```
`-EB` is already appended to `TOOLCHAIN_C_FLAGS` and `TOOLCHAIN_C_FLAGS` is added a compile options to Zephyr interface lib which is extracted here:
https://github.com/zephyrproject-rtos/zephyr/blob/4a69c1b8a8e92d6e77d139ce5dda41cc2012120e/cmake/modules/extensions.cmake#L5862-L5864

If this is not working as intended then we should found out why, instead of starting to duplicate all required flags.

@pillo79 have you noticed anything in this area not working as expected ?

laurenmurphyx64 replied:
@tejlmand @pillo79 zephyr_flags is fine, the command I wrote takes ${LLEXT_APPEND_FLAGS} as its only argument so it doesn't get those other flags. But when I try adding ${zephyr_filtered_flags}, or pulling  INTERFACE_COMPILE_OPTIONS from zephyr_interface, I get this error repeatedly:

```
Error evaluating generator expression:

    $<COMPILE_LANGUAGE:C>

  $<COMPILE_LANGUAGE:...> may only be used to specify include directories,
  compile definitions, compile options, and to evaluate components of the
  file(GENERATE) command.
```

I'm not great at CMake, but it seems to me like I can't use this particular generator expression in a custom command. I tried modifying the if(CONFIG_LLEXT_TYPE_ELF_OBJECT) to use an executable target so I could add the generator expression (${zephyr_filtered_flags}) as an argument to target_link_options. This pulls in every compiler flag to the link command.* Still, it works. I just had to add "-Wl,--no-gc-sections" to get qemu_leon3/leon3 to compile, as (I think?) picolibc.specs adds Wl,--gc-sections.

*It's worth noting that this is not how the other LLEXT_TYPEs do linking - they only pass ${LLEXT_APPEND_FLAGS}.

tejlmand commented on cmake/compiler/gcc/target_mips.cmake:
```diff
@@ -6,9 +6,11 @@
 else()
   list(APPEND TOOLCHAIN_C_FLAGS -EL)
   list(APPEND TOOLCHAIN_LD_FLAGS -EL)
+  set(LLEXT_APPEND_FLAGS -EL)
```
same comment as `-EB` applies.

tejlmand commented on cmake/compiler/host-gcc/target.cmake:
```diff
@@ -37,6 +37,16 @@ foreach(file_name include/stddef.h)
 
+if (CONFIG_64BIT)
+  set(LLEXT_APPEND_FLAGS
+    -m64
```
same comment as `-EB` applies.

laurenmurphyx64 replied:
Removed

tejlmand reviewed (changes requested):
just noticed that we are duplicating flags which ought to be extracted from already existing Zephyr compile options.

More explanation is needed or investigation as to why current code is not behaving as expected.

laurenmurphyx64 commented:
> > Do you want me to write a test that calls the loading functions up to llext_map_sections?
> 
> Based on original commit message I was under the impression that this PR was fixing arch which wasn't fully supported previously, and if that is the case then I assume that there are existing testcases here https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/subsys/llext/testcase.yaml to which those platforms can now be added to ensure they continue to work.

No, neither arch is supported yet. Sorry for the confusion :sweat:

mathieuchopstm commented on cmake/modules/extensions.cmake:
```diff
@@ -5965,34 +5982,35 @@ function(add_llext_target target_name)
   if(CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID)
     set(slid_inject_cmd
-      ${PYTHON_EXECUTABLE}
+      COMMAND ${PYTHON_EXECUTABLE}
```
Can you please split this change to a dedicated commit?

mathieuchopstm commented on cmake/modules/extensions.cmake:
```diff
@@ -5869,14 +5869,31 @@ function(add_llext_target target_name)
+
+    target_link_options(${llext_lib_target} PRIVATE
+      ${TOPT} ${ZEPHYR_BASE}/subsys/llext/llext_reorder_sections.ld
+      -Wl,--no-gc-sections ${zephyr_filtered_flags} ${LLEXT_APPEND_FLAGS}
```
Is `--no-gc-sections` toolchain-agnostic?

laurenmurphyx64 replied:
Probably not :sweat_smile: I added an if for the Zephyr toolchain.

laurenmurphyx64 committed:
llext: reorder TYPE_ELF_OBJECT with linker script

When processing ET_REL / object ELF files, apply a linker script to
reorder sections.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5874,9 +5874,31 @@ function(add_llext_target target_name)
+    set_ifndef(  TOPT -Wl,-T)  # Use this if the compiler driver doesn't set a value
+
+    set(no_gc_sections_flag)
+    if("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT})
```
we have the toolchain infrastructure in:
https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/toolchain
https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/compiler
https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/linker

where the toolchain / compiler / linker can describe flags as properties and then those can be fetched using property name instead.

Please rework.

laurenmurphyx64 commented on cmake/modules/extensions.cmake:
```diff
@@ -5874,9 +5874,31 @@ function(add_llext_target target_name)
+      $<TARGET_PROPERTY:linker,partial_linking>)
+    set_target_properties(${llext_lib_target} PROPERTIES
+      RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/llext
+      SUFFIX ${CMAKE_C_OUTPUT_EXTENSION})
```
Changed

laurenmurphyx64 commented on cmake/modules/extensions.cmake:
```diff
@@ -6001,21 +6001,21 @@ function(add_llext_target target_name)
   add_custom_command(
     OUTPUT ${llext_pkg_output}
-    COMMAND ${gnu_strip_for_mwdt_cmd}
+    ${gnu_strip_for_mwdt_cmd}
```
Changed back

pillo79 reviewed (commented):
Sorry to get back to this with so much delay. LGTM and thanks for your efforts, below is a non-blocking nit if you need to edit this.

However, I'd like to understand your use case a little bit better. It used to be that `OBJECT` did not involve the linker at all, and therefore was limited to one source file. With this change that is no longer true, so I guess it is now functionally equivalent to `RELOCATABLE`. What is the reason for not using that in your project in the first place?

Put it another way, can the linker script be added to the `RELOCATABLE` target instead (or if necessary, in addition to)? IMO, `OBJECT` is a relic of the very first LLEXT implementations and should really be deprecated/removed.

laurenmurphyx64 commented:
> However, I'd like to understand your use case a little bit better. It used to be that `OBJECT` did not involve the linker at all, and therefore was limited to one source file. With this change that is no longer true, so I guess it is now functionally equivalent to `RELOCATABLE`. What is the reason for not using that in your project in the first place?
> 
> Put it another way, can the linker script be added to the `RELOCATABLE` target instead (or if necessary, in addition to)? IMO, `OBJECT` is a relic of the very first LLEXT implementations and should really be deprecated/removed.

@pillo79 There's no particular use case attached to this work. I was just adding x86 and ARC support without a project in mind and noticed a couple test cases on those architectures failing due to sections being out of order, and then I saw platforms like qemu_leon and qemu_malta also compiled out of order for all test cases (once I got compilation working). But you have a point. I agree it should work on RELOCATABLE, too (I'll leave it up to y'all to decide when to remove OBJECT). Actually, I probably should've had that already :sweat_smile: 

Problem is, I'm running into [some issues](https://github.com/zephyrproject-rtos/zephyr/actions/runs/19832435365/job/56822026739?pr=93899) using the linker script with qemu_xtensa/dc233c and RELOCATABLE:
```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) llext.writable_relocatable on qemu_xtensa/dc233c failed (Testsuite failed)
INFO    - 2) llext.writable_relocatable_slid_linking on qemu_xtensa/dc233c failed (Testsuite failed)
INFO    - 3) sample.llext.modules.module_build on qemu_xtensa/dc233c failed (Timeout)
```

One is the linker inexplicably generating .rodata (and a couple other sections) with sh_addr 0x8, which I can solve by overriding in the linker script. The other seems to be an illegal instruction exception. I will work on this and get back to you.
